bundles/org.openhab.core.io.http.auth/src/main/java/org/openhab/core/io/http/auth/internal/CreateAPITokenPageServlet.java

```diff
@@ -0,0 +1,186 @@
+
+import java.io.IOException;
+import java.util.Map;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.auth.AuthenticationException;
+import org.openhab.core.auth.AuthenticationProvider;
+import org.openhab.core.auth.ManagedUser;
+import org.openhab.core.auth.User;
+import org.openhab.core.auth.UserRegistry;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.http.HttpService;
+import org.osgi.service.http.NamespaceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A servlet serving a page allowing users to create a new API token, after confirming their identity by signing in.
+ *
+ * @author Yannick Schaus - initial contribution
+ *
+ */
+@NonNullByDefault
+@Component(immediate = true)
+public class CreateAPITokenPageServlet extends AbstractAuthPageServlet {
+
+    private static final long serialVersionUID = 5340598701104679843L;
+
+    private final Logger logger = LoggerFactory.getLogger(CreateAPITokenPageServlet.class);
+
+    @Activate
+    public CreateAPITokenPageServlet(BundleContext bundleContext, @Reference HttpService httpService,
+            @Reference UserRegistry userRegistry, @Reference AuthenticationProvider authProvider) {
+        super(bundleContext, httpService, userRegistry, authProvider);
+        try {
+            httpService.registerServlet("/createApiToken", this, null, null);
+        } catch (NamespaceException | ServletException e) {
+            logger.error("Error during create API token page registration: {}", e.getMessage());
+        }
+    }
+
+    @Override
+    protected void doGet(@Nullable HttpServletRequest req, @Nullable HttpServletResponse resp)
+            throws ServletException, IOException {
+        if (req != null && resp != null) {
+            Map<String, String[]> params = req.getParameterMap();
+
+            try {
+                String message = "Create a new API token to authorize external services.";
+
+                // TODO: i18n
+                resp.setContentType("text/html;charset=UTF-8");
+                resp.getWriter().append(getPageBody(params, message, false));
+                resp.getWriter().close();
+            } catch (Exception e) {
+                resp.setContentType("text/plain;charset=UTF-8");
+                resp.getWriter().append(e.getMessage());
+                resp.getWriter().close();
+            }
+        }
+    }
+
+    @Override
+    protected void doPost(@Nullable HttpServletRequest req, @Nullable HttpServletResponse resp)
```

Comment:
This is against what was decided before. Otherwise using @NonNullByDefault({}) would also be possible for mandatory config fields or the config itself of set in the thing handler’s initialize(). The decision was to only use the @NonNullByDefault({})-annotation for injected ressources of the OSGi-framework.

Reply:
The agreed upon [guidelines](https://www.openhab.org/docs/developer/guidelines.html#null-annotations) don't mention anything about how to deal with annotating classes that implement methods from unannotated classes.

These issues also did not occur previously because we did not annotate classes with `@NonNullByDefault({})` if they were overriding methods from unannotated classes.

If there were EEAs for this dependency, these parameters would probably be analyzed to be `@NonNull` so in that case it makes more sense to use `@NonNullByDefault({})` which would prevent these unnecessary null checks.

Another case not mentioned in the guidelines is how to deal with generics. The `@NonNullByDefault({})` defined on classes is not used by the analysis for generic types. So in that case the only way to make them non-null is by annotating the generic types with `@NonNull`, see for example the [AbstractRegistry](https://github.com/openhab/openhab-core/blob/5d75bce5539aee321f263202f071304ffe2908b0/bundles/org.openhab.core/src/main/java/org/openhab/core/common/registry/AbstractRegistry.java#L58-L59).